File: rust/src/singly_linked_list.rs
pub struct ListNode<T> {
    value: T,
    next: Option<Box<ListNode<T>>>,
}

pub struct SinglyLinkedList<T> {
    length: usize,
    head: Option<Box<ListNode<T>>>,
}

impl<T> SinglyLinkedList<T> {
    pub fn new() -> SinglyLinkedList<T> {
        SinglyLinkedList {
            length: 0,
            head: None,
        }
    }

    pub fn prepend(&mut self, item: T) {
        let node = ListNode {
            value: item,
            next: self.head.take(),
        };
        self.head = Some(Box::new(node));
        self.length += 1;
    }

    pub fn insert_at(&mut self, item: T, idx: usize) -> Result<(), &'static str> {
        if idx > self.length {
            return Err("Index out of bounds");
        }

        if idx == 0 {
            self.prepend(item);
            return Ok(());
        }

        let mut current_node = self.head.as_mut();
        let mut prev_node: Option<&mut Box<ListNode<T>>> = None;
        let mut current_index = 0;

        while current_index < idx {
            prev_node = current_node;
            current_node = current_node.unwrap().next.as_mut();
            current_index += 1;
        }

        let new_node = ListNode {
            value: item,
            next: current_node.take().map(|node| Box::new(*node)),
        };
        prev_node.unwrap().next = Some(Box::new(new_node));
        self.length += 1;

        Ok(())
    }

    pub fn append(&mut self, item: T) {
        if self.head.is_none() {
            self.prepend(item);
            return;
        }

        let mut current_node = self.head.as_mut();

        while let Some(node) = current_node {
            if node.next.is_none() {
                node.next = Some(Box::new(ListNode { value: item, next: None }));
                self.length += 1;
                break;
            } else {
                current_node = node.next.as_mut();
            }
        }
    }

    pub fn remove(&mut self, item: T) -> Option<T>
    where
        T: PartialEq,
    {
        let mut current_node = self.head.as_mut();
        let mut prev_node: Option<&mut Box<ListNode<T>>> = None;

        if current_node.as_ref().map(|node| &node.value) == Some(&item) {
            self.head = self.head.unwrap().next;
            self.length -= 1;
            return Some(item);
        }

        while let Some(node) = current_node {
            if &node.value == &item {
                prev_node.unwrap().next = node.next.take();
                self.length -= 1;
                return Some(item);
            } else {
                prev_node = Some(node);
                current_node = node.next.as_mut();
            }
        }

        None
    }

    pub fn get(&self, idx: usize) -> Option<&T> {
        if idx >= self.length {
            return None;
        }

        let mut current_node = self.head.as_ref();
        let mut current_index = 0;

        while current_node.is_some() && current_index < idx {
            current_node = current_node.unwrap().next.as_ref();
            current_index += 1;
        }

        current_node.map(|node| &node.value)
    }

    pub fn remove_at(&mut self, idx: usize) -> Option<T> {
        if idx >= self.length {
            return None;
        }

        if idx == 0 {
            let value = self.head.take().unwrap().value;
            self.head = self.head.unwrap().next;
            self.length -= 1;
            return Some(value);
        }

        let mut current_node = self.head.as_mut();
        let mut prev_node: Option<&mut Box<ListNode<T>>> = None;
        let mut current_index = 0;

        while current_node.is_some() && current_index < idx {
            prev_node = current_node;
            current_node = current_node.unwrap().next.as_mut();
            current_index += 1;
        }

        if let Some(node) = current_node {
            prev_node.unwrap().next = node.next.take();
            self.length -= 1;
            return Some(node.value);
        }

        None
    }
}

File: java/src/SinglyLinkedList.java
package src;
// import src.ListNode;

// public class ListNode<T> {
//     T value;
//     ListNode<T> next;

//     public ListNode(T value) {
//         this.value = value;
//         this.next = null;
//     }
// }

public class SinglyLinkedList<T> {
    public int length;
    private ListNode<T> head;

    public SinglyLinkedList() {
        this.length = 0;
    }

    public void prepend(T item) {
        ListNode<T> node = new ListNode<>(item);
        node.next = this.head;
        this.head = node;
        this.length++;
    }

    public void insertAt(T item, int idx) {
        if (idx < 0 || idx > this.length) {
            throw new IllegalArgumentException("Index out of bounds");
        }

        if (idx == 0) {
            this.prepend(item);
            return;
        }

        ListNode<T> currentNode = this.head;
        ListNode<T> prevNode = null;
        int currentIndex = 0;

        while (currentIndex < idx) {
            prevNode = currentNode;
            currentNode = currentNode.next;
            currentIndex++;
        }

        ListNode<T> newNode = new ListNode<>(item);
        newNode.next = currentNode;
        prevNode.next = newNode;
        this.length++;
    }

    public void append(T item) {
        if (this.head == null) {
            this.prepend(item);
            return;
        }

        ListNode<T> currentNode = this.head;
        while (currentNode.next != null) {
            currentNode = currentNode.next;
        }

        currentNode.next = new ListNode<>(item);
        this.length++;
    }

    public T remove(T item) {
        if (this.head == null) {
            return null;
        }

        ListNode<T> currentNode = this.head;
        ListNode<T> prevNode = null;

        if (currentNode.value.equals(item)) {
            this.head = currentNode.next;
            this.length--;
            return item;
        }

        while (currentNode.next != null && !currentNode.value.equals(item)) {
            prevNode = currentNode;
            currentNode = currentNode.next;
        }

        if (currentNode.value.equals(item)) {
            prevNode.next = currentNode.next;
            this.length--;
            return item;
        }

        return null;
    }

    public T get(int idx) {
        if (idx < 0 || idx >= this.length) {
            return null;
        }

        ListNode<T> currentNode = this.head;
        int currentIndex = 0;

        while (currentNode != null && currentIndex < idx) {
            currentNode = currentNode.next;
            currentIndex++;
        }

        return currentNode != null ? currentNode.value : null;
    }

    public T removeAt(int idx) {
        if (idx < 0 || idx >= this.length) {
            return null;
        }

        if (idx == 0) {
            T value = this.head.value;
            this.head = this.head.next;
            this.length--;
            return value;
        }

        ListNode<T> currentNode = this.head;
        ListNode<T> prevNode = null;
        int currentIndex = 0;

        while (currentNode != null && currentIndex < idx) {
            prevNode = currentNode;
            currentNode = currentNode.next;
            currentIndex++;
        }

        if (currentNode != null) {
            prevNode.next = currentNode.next;
            this.length--;
            return currentNode.value;
        }

        return null;
    }
}
